From 4cd4af5520d7087f7677cbb7569feb5dc97e3984 Mon Sep 17 00:00:00 2001
From: Jeff Lin <jeffulin@google.com>
Date: Mon, 1 Jul 2024 15:37:35 +0800
Subject: [PATCH] Do not use boringssl in base::UnguessableToken

Introduced in r930012 (crrev.com/c/3343574).
Updated in r1230523 (crrev.com/c/5075453).
Updated in r1321254 (crrev.com/c/5669584).

Change-Id: I4035f5c236b9b2ca8b3c4f6296a36786cfe85ae7
---
 base/unguessable_token.cc |  7 ++-----
 base/unguessable_token.h  | 17 ++++++++---------
 2 files changed, 10 insertions(+), 14 deletions(-)

diff --git a/base/unguessable_token.cc b/base/unguessable_token.cc
index e996816597..7792a6b455 100644
--- a/base/unguessable_token.cc
+++ b/base/unguessable_token.cc
@@ -11,7 +11,6 @@
 #include "base/format_macros.h"
 #include "base/rand_util.h"
 #include "build/build_config.h"
-#include "third_party/boringssl/src/include/openssl/mem.h"
 
 namespace base {
 
@@ -54,10 +53,8 @@ std::optional<UnguessableToken> UnguessableToken::DeserializeFromString(
   return UnguessableToken(token.value());
 }
 
-bool operator==(const UnguessableToken& lhs, const UnguessableToken& rhs) {
-  auto bytes = lhs.token_.AsBytes();
-  auto other_bytes = rhs.token_.AsBytes();
-  return CRYPTO_memcmp(bytes.data(), other_bytes.data(), bytes.size()) == 0;
+bool UnguessableToken::operator==(const UnguessableToken& other) const {
+  return token_ == other.token_;
 }
 
 std::ostream& operator<<(std::ostream& out, const UnguessableToken& token) {
diff --git a/base/unguessable_token.h b/base/unguessable_token.h
index 76b7079b94..9b6a987597 100644
--- a/base/unguessable_token.h
+++ b/base/unguessable_token.h
@@ -8,7 +8,6 @@
 #include <stdint.h>
 #include <string.h>
 
-#include <compare>
 #include <iosfwd>
 #include <string_view>
 #include <utility>
@@ -107,12 +106,15 @@ class BASE_EXPORT UnguessableToken {
 
   span<const uint8_t, 16> AsBytes() const { return token_.AsBytes(); }
 
-  friend constexpr auto operator<=>(const UnguessableToken& lhs,
-                                    const UnguessableToken& rhs) = default;
+  constexpr bool operator<(const UnguessableToken& other) const {
+     return token_ < other.token_;
+   }
 
-  // operator== uses constant-time comparison for security where available.
-  friend BASE_EXPORT bool operator==(const UnguessableToken& lhs,
-                                     const UnguessableToken& rhs);
+  bool operator==(const UnguessableToken& other) const;
+
+  bool operator!=(const UnguessableToken& other) const {
+    return !(*this == other);
+  }
 
   template <typename H>
   friend H AbslHashValue(H h, const UnguessableToken& token) {
@@ -134,9 +136,6 @@ class BASE_EXPORT UnguessableToken {
   base::Token token_;
 };
 
-BASE_EXPORT bool operator==(const UnguessableToken& lhs,
-                            const UnguessableToken& rhs);
-
 BASE_EXPORT std::ostream& operator<<(std::ostream& out,
                                      const UnguessableToken& token);
 
-- 
2.50.0.727.gbf7dc18ff4-goog

